Add sensor for reading Austrian ZAMG weather conditions

mjl opened the pull request:
**Description:**

A sensor for fetching current weather data from the Austrian 
"Zentralanstalt für Meteorologie und Geodynamik" (zamg)

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1549

**Example entry for `configuration.yaml` (if applicable):**
```yaml
- sensor:
  - platform: zamg
    station_id: 11035
    monitored_conditions:
      - temperature
      - humidity

```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

Not yet: Documentation for configuration, supported conditions and station ids is in the module's docstring.

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. (Only py3.4)
  - [X] New dependencies have been added to the `REQUIREMENTS` variable (-> None).
  - [X] New dependencies are only imported inside functions that use them (-> None).
  - [X] New dependencies have been added to `requirements_all.txt1` (-> None)
  - [X] New files were added to `.coveragerc`.

Commit:
Add sensor for reading ZAMG weather conditions

Commit:
Add sensor for reading ZAMG weather conditions

Commit:
Add to coveragerc; Correct some doc style problems

Commit:
Add to coveragerc; Correct some doc style problems

Commit:
More doc fixes

Commit:
More doc fixes

Commit:
More doc fixes

Commit:
More doc fixes

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,283 @@
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# ------------------------------------------------------------------------
```
No need for a license statement. We ship the [License](https://github.com/home-assistant/home-assistant/blob/dev/LICENSE).

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,283 @@
+
+
+# ------------------------------------------------------------------------
+# pylint: disable=unused-argument
```
This is already [globally](https://github.com/home-assistant/home-assistant/blob/dev/pylintrc) disabled.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,273 @@
+
+
+# ------------------------------------------------------------------------
+class ZAMGWeather(WeatherEntity):
```
You need to decide if you want to create a platform for the `weather` component or a "classical" sensor.

balloob commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,283 @@
+})
+
+
+# ------------------------------------------------------------------------
```
Please remove these comment decorators.

balloob commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,283 @@
+DEFAULT_NAME = 'zamg'
+ATTRIBUTION = 'Data provided by ZAMG'
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=30)  # Only updates once per hour
```
This would be twice an hour.

balloob commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,283 @@
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_STATION_ID): cv.string,
```
You specify a list of available stations however you do not validate that the given station id is one of them.

balloob commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,283 @@
+        """Generic accessor for data."""
+        return self.data.get(variable, STATE_UNKNOWN)
+
+# ------------------------------------------------------------------------
```
Remove this

Commit:
Verify that the configured station id is actually one in the data set.

Don't warn about unknown stations, this cannot happen any more as the configuration parser now checks that.
This could still happen if the data set is incomplete though ...

Commit:
Verify that the configured station id is actually one in the data set.

Don't warn about unknown stations, this cannot happen any more as the configuration parser now checks that.
This could still happen if the data set is incomplete though ...

Commit:
Lose license and whatever.

Commit:
Lose license and whatever.

Commit:
Don't return UNKNOWN for unknown variables

Commit:
Don't return UNKNOWN for unknown variables

Commit:
Clean up imports

Commit:
Clean up imports

Commit:
Clarify comment on throttling interval

Commit:
Clarify comment on throttling interval

Commit:
Base zamg sensor on Entity, not WeatherEntity, and delete unused code

Commit:
Base zamg sensor on Entity, not WeatherEntity, and delete unused code

Commit:
Fix formatting nits from flake8

Commit:
Fix formatting nits from flake8

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,255 @@
+        """Return the state attributes."""
+        return {
+            ATTR_WEATHER_ATTRIBUTION: self.attribution,
+            "friendly_name": SENSOR_TYPES[self.variable][0],
```
There is `ATTR_FRIENDLY_NAME` available in `const.py`.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,255 @@
+    @property
+    def name(self):
+        """Build name of sensor."""
+        return "%s_%s" % (self.client_name, self.variable)
```
Use string formatting please. `'{} {}'.format(self.client_name, self.variable)`

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,255 @@
+    """
+
+    attribution = ATTRIBUTION
+    temperature_unit = TEMP_CELSIUS
```
I would like to suggest to avoid intermediate vars. Just use the constants directly.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,255 @@
+                                headers=self.API_HEADERS, timeout=15)
+        response.encoding = 'UTF8'
+        if response.status_code == 200:
+            content_type = response.headers.get('Content-Type', 'whatever')
```
This looks like a hack for the condition on Line 231. You could benefit of a guard clause here. Less indentation and simple code.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,255 @@
+                                   content_type)
+        else:
+            self._logger.error("API call returned with status %s",
+                               response.status_code)
```
I think that's worth to catch `requests.exceptions.RequestException` for the requests.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,255 @@
+
+    def __init__(self, logger, station_id):
+        """Initialize the probe."""
+        self._logger = logger
```
Looks like a duplicate. `_LOGGER` is already present.

mjl replied:
No, that's intentional, so one could test that sensor class by injecting a mock logger. I don't like to touch module or global scoped vars in multiple places if it can be avoided.

fabaff replied:
Would make sense if that's actually used. There are no tests available for this sensor.

fabaff reviewed (dismissed):
How about add `def unit_of_measurement(self):` to support units for the conditions

Commit:
Use ATTR_FRIENDLY_NAME, clean up imports, remove unnecessary indirection.

Commit:
Use ATTR_FRIENDLY_NAME, clean up imports, remove unnecessary indirection.

Commit:
Use {}.format() instead of "" %

Commit:
Use {}.format() instead of "" %

Commit:
Re-add unit of measurement that got lost somehow

Commit:
Re-add unit of measurement that got lost somehow

Commit:
Use guard clauses instead of if-matroshka.
Wrap requests.get() in try/except for RequestException.

Commit:
Use guard clauses instead of if-matroshka.
Wrap requests.get() in try/except for RequestException.

Commit:
Huh, how did this happen? White space corrections...

Commit:
Huh, how did this happen? White space corrections...

Commit:
Merge branch 'dev' of https://github.com/mjl/home-assistant into dev

* 'dev' of https://github.com/mjl/home-assistant:
  Huh, how did this happen? White space corrections...
  Use guard clauses instead of if-matroshka. Wrap requests.get() in try/except for RequestException.
  Re-add unit of measurement that got lost somehow
  Use {}.format() instead of "" %
  Use ATTR_FRIENDLY_NAME, clean up imports, remove unnecessary indirection.
  Fix formatting nits from flake8
  Base zamg sensor on Entity, not WeatherEntity, and delete unused code
  Clarify comment on throttling interval
  Clean up imports
  Verify that the configured station id is actually one in the data set.
  Don't return UNKNOWN for unknown variables
  Lose license and whatever.
  More doc fixes
  More doc fixes
  Add to coveragerc; Correct some doc style problems
  Add sensor for reading ZAMG weather conditions

Commit:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into dev

Commit:
Precipitation actually is a float, good it rained today

fabaff commented:
There are still comments which needed to be addressed.

mjl commented:
> Would make sense if that's actually used. There are no tests available for this sensor.

  I don't understand this comment. Why "would make sense if..."?

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
Logger needs no module visibility

fabaff commented:
You said "...so one could test that sensor class by injecting a mock logger" but you don't provide tests for the sensor.

mjl commented:
> You said "...so one could test that sensor class by injecting a mock logger" but you don't provide tests for the sensor.

  Ah I understand now. Just because something is not there (and I'm not sure it's worth the effort) doesn't mean one should write sloppy code. It doesn't hurt to do things right in the first place.

  Anyway, I believe we are rapidly approaching the point of diminishing results: are there any road blocks still open that I have overlooked other than minor personal style quibbles?

mjl commented:
Ah, it seems I picked the one other weather components as an how-to that names things with underscore: "PWS_" + self._condition.

Commit:
Do not name sensors with _ to be in line with the other weather sensor platforms.

Commit:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into dev

fabaff commented:
The `friendly_name` should be prefixed as well th be in sync with the existing platforms.

![weather](https://cloud.githubusercontent.com/assets/116184/20643806/4242937e-b422-11e6-8cec-ab1f37fa1a70.png)

Commit:
Remove manually set friendly_name

mjl commented:
> The friendly_name should be prefixed as well th be in sync with the existing platforms.

Ok, done.

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

# Conflicts:
#	.coveragerc

balloob commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,273 @@
+# ------------------------------------------------------------------------
```
Please remove this header.

Commit:
comment format police

balloob commented on .coveragerc:
```diff
@@ -310,7 +310,7 @@ omit =
     homeassistant/components/sensor/waqi.py
     homeassistant/components/sensor/xbox_live.py
     homeassistant/components/sensor/yweather.py
-    homeassistant/components/sensor/waqi.py
```
Why are you removing this ?

mjl replied:
Didn't, git merge conflict. Might I suggest a peek three lines higher?

balloob commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -0,0 +1,251 @@
+# coding=utf-8
```
Remove the whole utf and created by header please. The doc string should be the first part.

Commit:
Less comments

balloob committed:
Update zamg.py
